### ISSUE TITLE
[R-package] LGB.CV Record Evals 

### ISSUE THREAD

**Issue by lsemployeeoftheyear** [microsoft/LightGBM#5155]
## Description
In R I ran lgb.cv with the settings below. As I changed the parameters for nrounds and nfold, I found that the number of eval records under lgb.cv$record_evals$valid$kullback_leibler did not change. However, when I changed the num_tree variable under params, then the number of entries under lgb.cv$record_evals$valid$kullback_leibler went up. I also changed the nfold variable several times and found that it matches the number of boosters returned for the lgb.cv object. Additionally, I notice that for each tree there is one eval and one eval_err (scoring using kullback_leibler divergence).

1. What is the difference between "eval" and "eval_err" in lgb.cv$record_evals$valid$kullback_leibler? What exactly do these values represent?
2. Shouldn't the num_tree variable match the number of boosters and the nfold variable match the number of record_evals entries? Forgive me if I'm mistaken, but it looks to me like those variable have been reversed in the package somewhere.

## Reproducible example
library(lightgbm)
data(agaricus.train, package = 'lightgbm')
train = agaricus.train
etrain <- lgb.Dataset(data=train$data,
                      label=train$label)

Pars <- list(
  objective = "binary"
  , metric = "kullback_leibler"
  , min_data = 1L
  , learning_rate = 1.0
  , num_leaves = 31
  , num_tree = 100
  , predict_contrib = TRUE
  , is_unbalance= TRUE
)

model.cv <- lgb.cv(
  params = Pars
  , data = etrain
  , nrounds = 5
  , nfold = 10L
  , stratified = TRUE
)

model.cv$record_evals$valid$kullback_leibler

## Environment info

LightGBM version or commit hash:

Command(s) you used to install LightGBM

```shell
installed through cran within the last month
install.packages('lightgbm')
```

## Additional Comments

Thank you for your help :)

**Comment by jameslamb** [comment]
@lsemployeeoftheyear I'm sorry we haven't had a chance to get to this yet. Can you please explain why you closed it?

**Comment by jameslamb** [comment]
> Shouldn't the num_tree variable match the number of boosters and the nfold variable match the number of record_evals entries?

In LightGBM, a `Booster` refers to a single model object which contains a sequence of fitted trees. "booster" is not synonymous with "tree".

For a dataset with `N` rows, `lightgbm::lgb.cv()` will do the following:

1. randomly split the vector `1:N` into `nfold` non-overlapping integer vectors
2. for each of these sets of indices, hold out those rows as a validation set, and train for `num_trees` boosting rounds on the rest of the data
    - each of these produces 1 `Booster` object, which will have up to `num_trees` trees in it
    - after each iteration, calculate evaluation metrics by using that `Booster` to predict on the held-out data
3. return an `lgb.CVBooster` object, containing the following attributes:
    - `$boosters` = a list of each of `Booster` objects, one per fold
    - `$record_evals$valid` = a list of lists, one per metric, each of which has the following entries:
        - `$eval` = vector of length `num_trees`, where each element contains the mean of that evaluation metric as of that iteration, taken over each of the `nfold` Boosters
        - `$eval_err` = vector of length `num_trees` where each element contains the standard deviation of that evaluation metric as of that iteration, taken over each of the `nfold` Booster

Please also note that if you pass `num_trees` or any other aliases of `num_iterations` (https://lightgbm.readthedocs.io/en/latest/Parameters.html#num_iterations) via the `params` list, that value will override the keyword argument `nrounds`.

> What is the difference between "eval" and "eval_err" in `lgb.cv$record_evals$valid$kullback_leibler`? What exactly do these values represent?

`eval` contains the mean of a given metric over all folds.

`eval_err` contains the standard deviation of a given metric over all folds.

-----

@lsemployeeoftheyear I re-opened this after you closed it to be sure that it isn't closed until there's an answer to your questions. Otherwise, people with similar questions might find this from search engines and be disappointed to not find any information when they click into this issue.

I'm going to close this now that that answer has been provided. If you need any clarification on the answer above, please comment below and we can continue the discussion.

Thanks for using LightGBM!
